### PR TITLE
Add decoration parameter to DataTable and DataTableTheme

### LINKED ISSUE #65532: Add decoration parameter to DataTableThemeData to allow for custom background and border
Add decoration to the DataTableThemeData to allow for a background color and border set for a DataTable.

This is a follow up from https://github.com/flutter/flutter/pull/65383.

### PR DESCRIPTION
**Pull request by perclasson**
## Description

Add decoration parameter to DataTable and DataTableTheme.

Previously I used an `Ink` widget, which in turn uses an `LayoutBuilder`. This can cause issues when the parent widget is for example an `AlertDialog` and intrinsic sizing does not work. The work around is to use a transparent `Material` with a `Container` around it.

Example of a `DataTable` with a background and border set:

## Related Issues

Closes https://github.com/flutter/flutter/issues/65532 

## Tests

I added the following tests:

- DataTable renders with border and background decoration.
- Updated tests in `data_table_theme_data_test.dart` to assert on the decoration
- Had to add `first` to the `Container` finders in the `data_table_test.dart` as there's now one that wraps the whole `DataTable`

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I signed the [CLA].
- [X] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I updated/added relevant documentation (doc comments with `///`).
- [X] All existing and new tests are passing.
- [X] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [X] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [X] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/material/data_table.dart [hunk 1]
```diff
@@ -611,7 +619,7 @@ class DataTable extends StatelessWidget {
   /// Whether a border at the bottom of the table is displayed.
   ///
   /// By default, a border is not shown at the bottom to allow for a border
-  /// around the table.
+  /// around the table set with [DataTable.decoration].
```
Comment (guidezpl) [inline packages/flutter/lib/src/material/data_table.dart:622]: nit: can this be [decoration]?

**Review by guidezpl — APPROVED** [review]
```
  _    ___ _____ __  __ 
 | |  / __|_   _|  \/  |
 | |_| (_ | | | | |\/| |
 |____\___| |_| |_|  |_|
                        
```

**Review by JoseAlba — APPROVED** [review]
```
██╗      ██████╗ ████████╗███╗   ███╗
██║     ██╔════╝ ╚══██╔══╝████╗ ████║
██║     ██║  ███╗   ██║   ██╔████╔██║
██║     ██║   ██║   ██║   ██║╚██╔╝██║
███████╗╚██████╔╝   ██║   ██║ ╚═╝ ██║
╚══════╝ ╚═════╝    ╚═╝   ╚═╝     ╚═╝
```

File: packages/flutter/lib/src/material/data_table.dart (hunk 1 above)
Comment (perclasson) [reply packages/flutter/lib/src/material/data_table.dart:622]: Done!

**Review by HansMuller — COMMENTED** [review]
I may be missing something: expected to see tests that check DataTable with decoration borders with different widths and padding values; ensure that the table appears where it's expected.

File: packages/flutter/lib/src/material/data_table.dart [hunk 2]
```diff
@@ -611,7 +619,7 @@ class DataTable extends StatelessWidget {
   /// Whether a border at the bottom of the table is displayed.
   ///
   /// By default, a border is not shown at the bottom to allow for a border
-  /// around the table.
+  /// around the table set with [decoration].
```
[the remark this answers is not in the document]
Comment (perclasson) [reply packages/flutter/lib/src/material/data_table.dart:622]: Done!

File: packages/flutter/test/material/data_table_test.dart [hunk 3]
```diff
@@ -474,39 +474,39 @@ void main() {
       ),
     ));
     expect(tester.renderObject<RenderBox>(
-      find.widgetWithText(Container, 'Name')
+      find.widgetWithText(Container, 'Name').first
```
[the remark this answers is not in the document]
Comment (perclasson) [reply packages/flutter/test/material/data_table_test.dart:477]: I added the function `_findFirstContainerFor` now for this.

File: packages/flutter/test/material/data_table_theme_test.dart [hunk 4]
```diff
@@ -168,6 +178,7 @@ void main() {
     const double themeColumnSpacing = 3.0;
     const double themeDividerThickness = 4.0;
 
+    const BoxDecoration decoration = BoxDecoration(color: Color(0xfffffff0));
     final MaterialStateProperty<Color> dataRowColor = MaterialStateProperty.resolveWith<Color>(
```
[the remark this answers is not in the document]
Comment (perclasson) [reply packages/flutter/test/material/data_table_theme_test.dart:182]: Done!

**Comment by perclasson** [comment]
> I may be missing something: expected to see tests that check DataTable with decoration borders with different widths and padding values; ensure that the table appears where it's expected.

You're right, let me add a test for this.

File: packages/flutter/lib/src/material/data_table.dart [hunk 5]
```diff
@@ -975,9 +983,15 @@ class DataTable extends StatelessWidget {
       displayColumnIndex += 1;
     }
 
-    return Table(
-      columnWidths: tableColumns.asMap(),
-      children: tableRows,
+    return Container(
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/data_table.dart:986]: DecoratedBox would also work here.

File: packages/flutter/test/material/data_table_test.dart [hunk 6]
```diff
@@ -473,41 +473,35 @@ void main() {
...
       ),
     ));
-    expect(tester.renderObject<RenderBox>(
-      find.widgetWithText(Container, 'Name')
-    ).size.height, 56.0); // This is the header row height
-    expect(tester.renderObject<RenderBox>(
-      find.widgetWithText(Container, 'Frozen yogurt')
-    ).size.height, 48.0); // This is the data row height
+
+    // The finder matches with the Container of the cell content, as well as the
```
Comment (HansMuller) [inline packages/flutter/test/material/data_table_test.dart:477]: NICE

**Review by HansMuller — APPROVED** [review]
LGTM although the API doc needs to be expanded a bit.

File: packages/flutter/lib/src/material/data_table.dart [hunk 7]
```diff
@@ -473,6 +474,13 @@ class DataTable extends StatelessWidget {
   /// row is selectable.
   final ValueSetter<bool?>? onSelectAll;
 
+  /// {@template flutter.material.dataTable.decoration}
+  /// The background and border decoration for the table.
+  /// {@endtemplate}
+  ///
+  /// By default there is no decoration.
```
[the remark this answers is not in the document]
Comment (perclasson) [reply packages/flutter/lib/src/material/data_table.dart]: Done!

**Pull request merged**